File: App/Threads/ThreadComposeViewController.m
//  ThreadComposeViewController.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "ThreadComposeViewController.h"
#import "AwfulAppDelegate.h"
#import "AwfulForumsClient.h"
#import "AwfulForumTweaks.h"
#import "AwfulNewThreadFieldView.h"
#import "AwfulThreadPreviewViewController.h"
#import "AwfulThreadTagLoader.h"
#import "AwfulThreadTagPickerController.h"
#import "UINavigationItem+TwoLineTitle.h"
#import "Awful-Swift.h"

@interface ThreadComposeViewController () <AwfulThreadTagPickerControllerDelegate, UIViewControllerRestoration>

@property (strong, nonatomic) Thread *thread;

@property (strong, nonatomic) AwfulNewThreadFieldView *fieldView;
@property (strong, nonatomic) AwfulThreadTagPickerController *threadTagPicker;
@property (strong, nonatomic) ThreadTag *threadTag;
@property (strong, nonatomic) ThreadTag *secondaryThreadTag;

@property (copy, nonatomic) void (^onAppearBlock)(void);

@property (copy, nonatomic) NSArray *availableThreadTags;
@property (copy, nonatomic) NSArray *availableSecondaryThreadTags;
@property (assign, nonatomic) BOOL updatingThreadTags;

@end

@implementation ThreadComposeViewController

- (instancetype)initWithForum:(Forum *)forum
{
    if ((self = [super initWithNibName:nil bundle:nil])) {
        _forum = forum;
        self.title = DefaultTitle;
        self.submitButtonItem.title = @"Preview";
        self.restorationClass = self.class;
        [self updateTweaks];
    }
    return self;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    NSAssert(nil, @"Use -initWithForum: instead");
    return [self initWithForum:nil];
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    NSAssert(nil, @"Use -initWithForum: instead");
    return [self initWithForum:nil];
}

static NSString * const DefaultTitle = @"New Thread";

- (void)setTitle:(NSString *)title
{
    [super setTitle:title];
    self.navigationItem.titleLabel.text = title;
}

- (void)setThreadTag:(ThreadTag *)threadTag
{
    _threadTag = threadTag;
    [self updateThreadTagButtonImage];
}

- (void)setSecondaryThreadTag:(ThreadTag *)secondaryThreadTag
{
    _secondaryThreadTag = secondaryThreadTag;
    [self updateThreadTagButtonImage];
}

- (Theme *)theme
{
    return [Theme currentThemeForForum:self.forum];
}

- (void)loadView
{
    [super loadView];
    self.customView = self.fieldView;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self updateThreadTagButtonImage];
    [self updateAvailableThreadTagsIfNecessary];
}

- (void)updateAvailableThreadTagsIfNecessary
{
    if (self.availableThreadTags.count > 0 || self.updatingThreadTags) return;
    
    self.updatingThreadTags = YES;
    __weak __typeof__(self) weakSelf = self;
    [[AwfulForumsClient client] listAvailablePostIconsForForumWithID:self.forum.forumID andThen:^(NSError *error, AwfulForm *form) {
        __typeof__(self) self = weakSelf;
        self.availableThreadTags = form.threadTags;
        self.availableSecondaryThreadTags = form.secondaryThreadTags;
        self.updatingThreadTags = NO;
    }];
}

- (void)themeDidChange
{
    [super themeDidChange];
    self.fieldView.subjectField.textField.textColor = self.textView.textColor;
    self.fieldView.subjectField.textField.keyboardAppearance = self.textView.keyboardAppearance;
    
    NSDictionary *styleAttrs = @{NSForegroundColorAttributeName: self.theme[@"placeholderTextColor"]};
    NSAttributedString *themedString = [[NSAttributedString alloc] initWithString:@"Subject" attributes:styleAttrs];
    self.fieldView.subjectField.textField.attributedPlaceholder = themedString;
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self updateAvailableThreadTagsIfNecessary];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    if (self.onAppearBlock) {
        self.onAppearBlock();
        self.onAppearBlock = nil;
    }
}

- (void)updateTweaks
{
	AwfulForumTweaks *tweaks = [AwfulForumTweaks tweaksWithForumID:self.forum.forumID];
	
	//Apply autocorrection tweaks to subject field
	self.fieldView.subjectField.textField.autocapitalizationType = tweaks.autocapitalizationType;
    self.fieldView.subjectField.textField.autocorrectionType = tweaks.autocorrectionType;
    self.fieldView.subjectField.textField.spellCheckingType = tweaks.spellCheckingType;
	
	//Apply autocorrection tweaks to main text view
	self.textView.autocapitalizationType = tweaks.autocapitalizationType;
    self.textView.autocorrectionType = tweaks.autocorrectionType;
    self.textView.spellCheckingType = tweaks.spellCheckingType;
}

- (AwfulNewThreadFieldView *)fieldView
{
    if (_fieldView) return _fieldView;
    _fieldView = [[AwfulNewThreadFieldView alloc] initWithFrame:CGRectMake(0, 0, 0, 45)];
    _fieldView.subjectField.label.textColor = [UIColor grayColor];
    [_fieldView.threadTagButton addTarget:self
                                   action:@selector(didTapThreadTagButton:)
                         forControlEvents:UIControlEventTouchUpInside];
    [_fieldView.subjectField.textField addTarget:self
                                          action:@selector(subjectFieldDidChange:)
                                forControlEvents:UIControlEventEditingChanged];
    return _fieldView;
}

- (void)updateThreadTagButtonImage
{
    UIImage *image;
    if (self.threadTag) {
        image = [AwfulThreadTagLoader imageNamed:self.threadTag.imageName];
    } else {
        image = [AwfulThreadTagLoader unsetThreadTagImage];
    }
    [self.fieldView.threadTagButton setImage:image forState:UIControlStateNormal];
    if (self.secondaryThreadTag) {
        ThreadTag *tag = self.secondaryThreadTag;
        image = [AwfulThreadTagLoader imageNamed:tag.imageName];
    } else {
        image = nil;
    }
    self.fieldView.threadTagButton.secondaryTagImage = image;
}

- (void)didTapThreadTagButton:(UIButton *)button
{
    // TODO better handle waiting for the available thread tags to download
    if (self.availableThreadTags.count == 0) return;
    
    NSString *selectedImageName = self.threadTag.imageName ?: AwfulThreadTagLoaderEmptyThreadTagImageName;
    [self.threadTagPicker selectImageName:selectedImageName];
    if (self.availableSecondaryThreadTags.count > 0) {
        NSString *selectedSecondaryImageName = self.secondaryThreadTag.imageName ?: [self.availableSecondaryThreadTags[0] imageName];
        [self.threadTagPicker selectSecondaryImageName:selectedSecondaryImageName];
    }
    [self.threadTagPicker presentFromView:button];
    
    // HACK: Calling -endEditing: once doesn't work if the Subject field is selected. But twice works. I assume this is some weirdness from adding a text field as a subview to a text view.
    [self.view endEditing:YES];
    [self.view endEditing:YES];
}

- (AwfulThreadTagPickerController *)threadTagPicker
{
    if (_threadTagPicker) return _threadTagPicker;
    if (self.availableThreadTags.count == 0) return nil;
    
    NSMutableArray *imageNames = [NSMutableArray arrayWithObject:AwfulThreadTagLoaderEmptyThreadTagImageName];
    [imageNames addObjectsFromArray:[self.availableThreadTags valueForKey:@"imageName"]];
    NSArray *secondaryImageNames = [self.availableSecondaryThreadTags valueForKey:@"imageName"];
    _threadTagPicker = [[AwfulThreadTagPickerController alloc] initWithImageNames:imageNames secondaryImageNames:secondaryImageNames];
    _threadTagPicker.delegate = self;
    _threadTagPicker.title = @"Choose Thread Tag";
    if (self.availableSecondaryThreadTags.count > 0) {
        _threadTagPicker.navigationItem.rightBarButtonItem = _threadTagPicker.doneButtonItem;
    } else {
        _threadTagPicker.navigationItem.leftBarButtonItem = _threadTagPicker.cancelButtonItem;
    }
    return _threadTagPicker;
}

- (void)subjectFieldDidChange:(UITextField *)subjectTextField
{
    self.title = subjectTextField.text.length > 0 ? subjectTextField.text : DefaultTitle;
    [self updateSubmitButtonItem];
}

- (BOOL)canSubmitComposition
{
    return ([super canSubmitComposition] && self.fieldView.subjectField.textField.text.length > 0 && self.threadTag);
}

- (void)shouldSubmitHandler:(void(^)(BOOL ok))handler
{
    AwfulThreadPreviewViewController *preview = [[AwfulThreadPreviewViewController alloc] initWithForum:self.forum
                                                                                                subject:self.fieldView.subjectField.textField.text
                                                                                              threadTag:self.threadTag
                                                                                     secondaryThreadTag:self.secondaryThreadTag
                                                                                                 BBcode:self.textView.attributedText];
    preview.submitBlock = ^{ handler(YES); };
    self.onAppearBlock = ^{ handler(NO); };
    [self.navigationController pushViewController:preview animated:YES];
}

- (NSString *)submissionInProgressTitle
{
    return @"Posting…";
}

- (void)submitComposition:(NSString *)composition completionHandler:(void(^)(BOOL success))completionHandler
{
    __weak __typeof__(self) weakSelf = self;
    [[AwfulForumsClient client] postThreadInForum:self.forum
                                      withSubject:self.fieldView.subjectField.textField.text
                                        threadTag:self.threadTag
                                     secondaryTag:self.secondaryThreadTag
                                           BBcode:composition
                                          andThen:^(NSError *error, Thread *thread)
    {
        __typeof__(self) self = weakSelf;
        if (error) {
            UIAlertController *alert = [[UIAlertController alloc] initWithTitle:@"Network Error" error:error handler:^(UIAlertAction *action) {
                completionHandler(NO);
            }];
            [self presentViewController:alert animated:YES completion:nil];
        } else {
            self.thread = thread;
            completionHandler(YES);
        }
    }];
}

#pragma mark - AwfulThreadTagPickerControllerDelegate

- (void)threadTagPicker:(AwfulThreadTagPickerController *)picker didSelectImageName:(NSString *)imageName
{
    if ([imageName isEqualToString:AwfulThreadTagLoaderEmptyThreadTagImageName]) {
        self.threadTag = nil;
    } else {
        [self.availableThreadTags enumerateObjectsUsingBlock:^(ThreadTag *threadTag, NSUInteger i, BOOL *stop) {
            if ([threadTag.imageName isEqualToString:imageName]) {
                self.threadTag = threadTag;
                *stop = YES;
            }
        }];
    }
    if (self.availableSecondaryThreadTags.count == 0) {
        [picker dismiss];
        [self focusInitialFirstResponder];
    }
}

- (void)threadTagPicker:(AwfulThreadTagPickerController *)picker didSelectSecondaryImageName:(NSString *)secondaryImageName
{
    [self.availableSecondaryThreadTags enumerateObjectsUsingBlock:^(ThreadTag *secondaryThreadTag, NSUInteger i, BOOL *stop) {
        if ([secondaryThreadTag.imageName isEqualToString:secondaryImageName]) {
            self.secondaryThreadTag = secondaryThreadTag;
            *stop = YES;
        }
    }];
}

- (void)threadTagPickerDidDismiss:(AwfulThreadTagPickerController *)picker
{
    [self focusInitialFirstResponder];
}

#pragma mark - State preservation and restoration

+ (UIViewController *)viewControllerWithRestorationIdentifierPath:(NSArray *)identifierComponents coder:(NSCoder *)coder
{
    ForumKey *forumKey = [coder decodeObjectForKey:ForumKeyKey];
    if (!forumKey) {
        // AwfulObjectKey was introduced in Awful 3.2.
        NSString *forumID = [coder decodeObjectForKey:obsolete_ForumIDKey];
        forumKey = [[ForumKey alloc] initWithForumID:forumID];
    }
    Forum *forum = [Forum objectForKey:forumKey inManagedObjectContext:[AwfulAppDelegate instance].managedObjectContext];
    ThreadComposeViewController *newThreadViewController = [[ThreadComposeViewController alloc] initWithForum:forum];
    newThreadViewController.restorationIdentifier = identifierComponents.lastObject;
    return newThreadViewController;
}

- (void)encodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super encodeRestorableStateWithCoder:coder];
    [coder encodeObject:self.forum.objectKey forKey:ForumKeyKey];
    [coder encodeObject:self.fieldView.subjectField.textField.text forKey:SubjectKey];
    [coder encodeObject:self.threadTag.objectKey forKey:ThreadTagKeyKey];
    [coder encodeObject:self.secondaryThreadTag.objectKey forKey:SecondaryThreadTagKeyKey];
}

- (void)decodeRestorableStateWithCoder:(NSCoder *)coder
{
    self.fieldView.subjectField.textField.text = [coder decodeObjectForKey:SubjectKey];
    // AwfulObjectKey was introduced in Awful 3.2.
    ThreadTagKey *tagKey = [coder decodeObjectForKey:ThreadTagKeyKey];
    if (!tagKey) {
        NSString *threadTagImageName = [coder decodeObjectForKey:obsolete_ThreadTagImageNameKey];
        if (threadTagImageName) {
            tagKey = [[ThreadTagKey alloc] initWithImageName:threadTagImageName threadTagID:nil];
        }
    }
    if (tagKey) {
        self.threadTag = [ThreadTag objectForKey:tagKey inManagedObjectContext:self.forum.managedObjectContext];
    }
    ThreadTagKey *secondaryTagKey = [coder decodeObjectForKey:SecondaryThreadTagKeyKey];
    if (!secondaryTagKey) {
        NSString *secondaryTagImageName = [coder decodeObjectForKey:obsolete_SecondaryThreadTagImageNameKey];
        if (secondaryTagImageName) {
            secondaryTagKey = [[ThreadTagKey alloc] initWithImageName:secondaryTagImageName threadTagID:nil];
        }
    }
    if (secondaryTagKey) {
        self.secondaryThreadTag = [ThreadTag objectForKey:secondaryTagKey inManagedObjectContext:self.forum.managedObjectContext];
    }
    
    [super decodeRestorableStateWithCoder:coder];
}

static NSString * const ForumKeyKey = @"ForumKey";
static NSString * const obsolete_ForumIDKey = @"AwfulForumID";
static NSString * const SubjectKey = @"AwfulSubject";
static NSString * const ThreadTagKeyKey = @"ThreadTagKey";
static NSString * const obsolete_ThreadTagImageNameKey = @"AwfulThreadTagImageName";
static NSString * const SecondaryThreadTagKeyKey = @"SecondaryThreadTagKey";
static NSString * const obsolete_SecondaryThreadTagImageNameKey = @"AwfulSecondaryThreadTagImageName";

@end

File: App/Threads/ThreadComposeViewController.swift
//  ThreadComposeViewController.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import AwfulCore

/// For writing the OP of a new thread.
final class ThreadComposeViewController: ComposeTextViewController {
    /// The newly-posted thread.
    fileprivate(set) var thread: AwfulThread?
    fileprivate let forum: Forum
    fileprivate var threadTag: ThreadTag? {
        didSet { updateThreadTagButtonImage() }
    }
    fileprivate var secondaryThreadTag: ThreadTag? {
        didSet { updateThreadTagButtonImage() }
    }
    fileprivate var fieldView: NewThreadFieldView!
    fileprivate var availableThreadTags: [ThreadTag]?
    fileprivate var availableSecondaryThreadTags: [ThreadTag]?
    fileprivate var updatingThreadTags = false
    fileprivate var onAppearBlock: (() -> Void)?
    fileprivate var threadTagPicker: ThreadTagPickerViewController?
    
    /// - parameter forum: The forum in which the new thread is posted.
    init(forum: Forum) {
        self.forum = forum
        super.init(nibName: nil, bundle: nil)
        
        title = defaultTitle
        submitButtonItem.title = "Preview"
        restorationClass = type(of: self)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override var title: String? {
        didSet { navigationItem.titleLabel.text = title }
    }
    
    override func loadView() {
        super.loadView()
        
        fieldView = NewThreadFieldView(frame: CGRect(x: 0, y: 0, width: 0, height: 45))
        fieldView.subjectField.label.textColor = .gray
        fieldView.threadTagButton.addTarget(self, action: #selector(didTapThreadTagButton), for: .touchUpInside)
        fieldView.subjectField.textField.addTarget(self, action: #selector(subjectFieldDidChange), for: .editingChanged)
        customView = fieldView
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        updateTweaks()
        updateThreadTagButtonImage()
        updateAvailableThreadTagsIfNecessary()
    }
    
    override var theme: Theme {
        return Theme.currentThemeForForum(forum: forum)
    }
    
    override func themeDidChange() {
        super.themeDidChange()
        
        fieldView.subjectField.textField.textColor = textView.textColor
        fieldView.subjectField.textField.keyboardAppearance = textView.keyboardAppearance
        
        let attributes = [NSForegroundColorAttributeName: theme["placeholderTextColor"] ?? .gray]
        let themedString = NSAttributedString(string: "Subject", attributes: attributes)
        fieldView.subjectField.textField.attributedPlaceholder = themedString
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        updateAvailableThreadTagsIfNecessary()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        onAppearBlock?()
        onAppearBlock = nil
    }
    
    fileprivate func updateTweaks() {
        guard let tweaks = ForumTweaks(forumID: forum.forumID) else { return }
        fieldView.subjectField.textField.autocapitalizationType = tweaks.autocapitalizationType
        fieldView.subjectField.textField.autocorrectionType = tweaks.autocorrectionType
        fieldView.subjectField.textField.spellCheckingType = tweaks.spellCheckingType
        
        textView.autocapitalizationType = tweaks.autocapitalizationType
        textView.autocorrectionType = tweaks.autocorrectionType
        textView.spellCheckingType = tweaks.spellCheckingType
    }
    
    fileprivate func updateThreadTagButtonImage() {
        let image: UIImage?
        if let imageName = threadTag?.imageName {
            image = ThreadTagLoader.imageNamed(imageName)
        } else {
            image = ThreadTagLoader.unsetThreadTagImage
        }
        fieldView.threadTagButton.setImage(image, for: UIControlState())
        
        let secondaryImage: UIImage?
        if let imageName = secondaryThreadTag?.imageName {
            secondaryImage = ThreadTagLoader.imageNamed(imageName)
        } else {
            secondaryImage = nil
        }
        fieldView.threadTagButton.secondaryTagImage = secondaryImage
    }
    
    @objc fileprivate func didTapThreadTagButton(_ sender: UIButton) {
        guard let picker = threadTagPicker else { return }
        
        let selectedImageName = threadTag?.imageName ?? ThreadTagLoader.emptyThreadTagImageName
        picker.selectImageName(selectedImageName)
        
        if let
            secondaryTags = availableSecondaryThreadTags,
            let secondaryImageName = secondaryThreadTag?.imageName ?? secondaryTags.first?.imageName
        {
            picker.selectSecondaryImageName(secondaryImageName)
        }
        
        picker.present(fromView: sender)
        
        // HACK: Calling -endEditing: once doesn't work if the Subject field is selected. But twice works. I assume this is some weirdness from adding a text field as a subview to a text view.
        view.endEditing(true)
        view.endEditing(true)
    }
    
    @objc fileprivate func subjectFieldDidChange(_ sender: UITextField) {
        if let text = sender.text , !text.isEmpty {
            title = text
        } else {
            title = defaultTitle
        }
        
        updateSubmitButtonItem()
    }
    
    fileprivate func updateAvailableThreadTagsIfNecessary() {
        guard availableThreadTags == nil && !updatingThreadTags else { return }
        
        updatingThreadTags = true
        _ = ForumsClient.shared.listAvailablePostIcons(inForumIdentifiedBy: forum.forumID)
            .then { [weak self] (form) -> Void in
                guard let sself = self else { return }
                sself.updatingThreadTags = false
                sself.availableThreadTags = form.threadTags
                sself.availableSecondaryThreadTags = form.secondaryThreadTags
                guard let tags = sself.availableThreadTags else { return }

                let imageNames = [ThreadTagLoader.emptyThreadTagImageName] + tags.flatMap { $0.imageName }
                let secondaryImageNames = sself.availableSecondaryThreadTags?.flatMap { $0.imageName }
                let picker = ThreadTagPickerViewController(imageNames: imageNames, secondaryImageNames: secondaryImageNames)
                sself.threadTagPicker = picker
                picker.delegate = sself
                picker.title = "Choose Thread Tag"
                if secondaryImageNames?.isEmpty == false {
                    picker.navigationItem.rightBarButtonItem = picker.doneButtonItem
                } else {
                    picker.navigationItem.leftBarButtonItem = picker.cancelButtonItem
                }
            }
            .catch { [weak self] (error) -> Void in
                guard let sself = self else { return }
                sself.updatingThreadTags = false
                sself.availableThreadTags = nil
                sself.availableSecondaryThreadTags = nil
        }
    }
    
    override var canSubmitComposition: Bool {
        guard super.canSubmitComposition else { return false }
        guard fieldView.subjectField.textField.text?.isEmpty == false else { return false }
        return threadTag != nil
    }
    
    override func shouldSubmit(_ handler: @escaping (Bool) -> Void) {
        guard let
            subject = fieldView.subjectField.textField.text,
            let threadTag = threadTag
            else { return handler(false) }
        let preview = ThreadPreviewViewController(forum: forum, subject: subject, threadTag: threadTag, secondaryThreadTag: secondaryThreadTag, BBcode: textView.attributedText)
        preview.submitBlock = { handler(true) }
        onAppearBlock = { handler(false) }
        navigationController?.pushViewController(preview, animated: true)
    }
    
    override var submissionInProgressTitle: String {
        return "Posting…"
    }
    
    override func submit(_ composition: String, completion: @escaping (Bool) -> Void) {
        guard
            let subject = fieldView.subjectField.textField.text,
            let threadTag = threadTag
            else { return completion(false) }
        
        _ = ForumsClient.shared.postThread(in: forum, subject: subject, threadTag: threadTag, secondaryTag: secondaryThreadTag, bbcode: composition)
            .then { [weak self] (thread) -> Void in
                self?.thread = thread
                completion(true)
            }
            .catch { [weak self] (error) -> Void in
                let alert = UIAlertController(title: "Network Error", error: error, handler: { (action) in
                    completion(false)
                })
                self?.present(alert, animated: true, completion: nil)
        }
    }
    
    override func encodeRestorableState(with coder: NSCoder) {
        super.encodeRestorableState(with: coder)
        
        coder.encode(forum.objectKey, forKey: Keys.ForumKey.rawValue)
        coder.encode(fieldView.subjectField.textField.text, forKey: Keys.SubjectKey.rawValue)
        coder.encode(threadTag?.objectKey, forKey: Keys.ThreadTagKey.rawValue)
        coder.encode(secondaryThreadTag?.objectKey, forKey: Keys.SecondaryThreadTagKey.rawValue)
    }
    
    override func decodeRestorableState(with coder: NSCoder) {
        fieldView.subjectField.textField.text = coder.decodeObject(forKey: Keys.SubjectKey.rawValue) as? String
        
        if let tagKey = coder.decodeObject(forKey: Keys.ThreadTagKey.rawValue) as? ThreadTagKey {
            threadTag = ThreadTag.objectForKey(objectKey: tagKey, inManagedObjectContext: forum.managedObjectContext!) as? ThreadTag
        }
        
        if let secondaryTagKey = coder.decodeObject(forKey: Keys.SecondaryThreadTagKey.rawValue) as? ThreadTagKey {
            secondaryThreadTag = ThreadTag.objectForKey(objectKey: secondaryTagKey, inManagedObjectContext: forum.managedObjectContext!) as? ThreadTag
        }
        
        super.decodeRestorableState(with: coder)
    }
}

extension ThreadComposeViewController: ThreadTagPickerViewControllerDelegate {
    func threadTagPicker(_ picker: ThreadTagPickerViewController, didSelectImageName imageName: String) {
        if imageName == ThreadTagLoader.emptyThreadTagImageName {
            threadTag = nil
        } else if let
            threadTags = availableThreadTags,
            let i = threadTags.index(where: { $0.imageName == imageName})
        {
            threadTag = threadTags[i]
        }
        
        if availableSecondaryThreadTags?.isEmpty ?? true {
            picker.dismiss()
            focusInitialFirstResponder()
        }
    }
    
    func threadTagPicker(_ picker: ThreadTagPickerViewController, didSelectSecondaryImageName imageName: String) {
        if let
            tags = availableSecondaryThreadTags,
            let i = tags.index(where: { $0.imageName == imageName })
        {
            secondaryThreadTag = tags[i]
        }
    }
    
    func threadTagPickerDidDismiss(_ picker: ThreadTagPickerViewController) {
        focusInitialFirstResponder()
    }
}

extension ThreadComposeViewController: UIViewControllerRestoration {
    static func viewController(withRestorationIdentifierPath identifierComponents: [Any], coder: NSCoder) -> UIViewController? {
        guard let
            forumKey = coder.decodeObject(forKey: Keys.ForumKey.rawValue) as? ForumKey,
            let forum = Forum.objectForKey(objectKey: forumKey, inManagedObjectContext: AppDelegate.instance.managedObjectContext) as? Forum
            else { return nil }
        
        let composeViewController = ThreadComposeViewController(forum: forum)
        composeViewController.restorationIdentifier = identifierComponents.last as? String
        return composeViewController
    }
}

private let defaultTitle = "New Thread"

private enum Keys: String {
    case ForumKey
    case SubjectKey = "AwfulSubject"
    case ThreadTagKey
    case SecondaryThreadTagKey
}
